shared/editor/plugins/Mermaid.ts

```diff
@@ -0,0 +1,175 @@
+import { Node } from "prosemirror-model";
+import { Plugin, PluginKey, Transaction } from "prosemirror-state";
+import { findBlockNodes } from "prosemirror-utils";
+import { Decoration, DecorationSet } from "prosemirror-view";
+
+const positionDiagramIdMap: Record<number, number> = {};
+
+type MermaidState = {
+  lastDiagramIdGenerated: number;
+  decorationSet: DecorationSet;
+  diagramVisibility: Record<number, boolean>;
+};
+
+function getNewState({
+  doc,
+  name,
+  pluginState,
+}: {
+  doc: Node;
+  name: string;
+  pluginState: MermaidState;
+}) {
+  const decorations: Decoration[] = [];
+  const blocks: { node: Node; pos: number }[] = findBlockNodes(doc).filter(
+    (item) => item.node.type.name === name
+  );
+
+  blocks.forEach((block) => {
+    const language = block.node.attrs.language;
+    if (!language || language !== "mermaidjs") {
+      return;
+    }
+
+    let diagramId: number = positionDiagramIdMap[block.pos];
```

Comment:
So here, I think you can use the `block.node + block.node.nodeSize` to get the position of the decoration, from which you can extract the previous diagramId?

Reply:
Good idea! But how can I access the DOM from here? I dont know how to get an instance of the EditorView at that place, which would allow me to use view.domAtPos to get the diagramId

Reply:
You don't need to, you can get it off the decoration directly I think?

Reply:
The old state is available, but you need to pass it into getNewState

Reply:
Ah, you are completely right. Thanks!